src/oneD/refine.cpp

```diff
@@ -77,153 +80,152 @@ int Refiner::analyze(size_t n, const double* z, const double* x)
+            // consist of small fluctuations around a constant value.
+            if (valMax - valMin > m_min_range*valMagnitude) {
+                // maximum allowable difference in value between adjacent points. Based
+                // on the global min and max values of the component over the domain.
+                double max_change = m_slope*(valMax - valMin);
                 for (size_t j = 0; j < n-1; j++) {
-                    double r = fabs(v[j+1] - v[j])/dmax;
-                    if (r > 1.0 && dz[j] >= 2 * m_gridmin) {
-                        m_loc.insert(j);
-                        m_c.insert(name);
+                    double ratio = fabs(val[j+1] - val[j]) / (max_change + m_thresh);
+                    if (ratio > 1.0 && dz[j] >= 2 * m_gridmin) {
+                        m_insertPts.insert(j);
+                        m_componentNames.insert(name);
                     }
-                    if (r >= m_prune) {
-                        m_keep[j] = 1;
-                        m_keep[j+1] = 1;
-                    } else if (m_keep[j] == 0) {
-                        m_keep[j] = -1;
+                    if (ratio >= m_prune) {
+                        m_keep[j] = KEEP;
+                        m_keep[j+1] = KEEP;
+                    } else if (m_keep[j] == UNSET) {
+                        m_keep[j] = REMOVE;
                     }
                 }
             }
 
-            // refine based on the slope of component i only if the
-            // range of s is greater than a fraction 'min_range' of max
-            // |s|. This eliminates components that consist of small
-            // fluctuations on a constant slope background.
-            if ((smax - smin) > m_min_range*ss) {
-                // maximum allowable difference in slope between
-                // adjacent points.
-                double dmax = m_curve*(smax - smin);
+            // refine based on the slope of component i only if the range of s is
+            // greater than a fraction 'min_range' of max|s|. This eliminates
+            // components that consist of small fluctuations on a constant slope
+            // background.
+            if (slopeMax - slopeMin > m_min_range*slopeMagnitude) {
+                // maximum allowable difference in slope between adjacent points.
+                double max_change = m_curve*(slopeMax - slopeMin);
                 for (size_t j = 0; j < n-2; j++) {
-                    double r = fabs(s[j+1] - s[j]) / (dmax + m_thresh/dz[j]);
-                    if (r > 1.0 && dz[j] >= 2 * m_gridmin &&
-                            dz[j+1] >= 2 * m_gridmin) {
-                        m_c.insert(name);
-                        m_loc.insert(j);
-                        m_loc.insert(j+1);
+                    // Using the solution component absolute tolerance (m_thresh),
+                    // a smallest allowable slope can be estimated for an interval dz
+                    // as m_thresh/dz.
```

Comment:
This still reads strangely to me. There is no "smallest allowable slope" -- it is perfectly fine for a component to have zero slope. The point of adding this term is to avoid adding a point when the absolute change in the slope is very small, even if the change in slope is large in a relative sense, which can occur if `max_change` is small.

Reply:
I agree that it's not clear.  Is it that the outer check `slopeMax - slopeMin > m_min_range*slopeMagnitude` is only checking for noise around a constant non-zero-ish value. And then inside the if-statement the `m_thresh/dz` is catching that other case where the max change over the entire domain is very close to zero? It's the description of why we need to perform a `m_thresh/dz` that's tripping me up. An absolute tolerance that is dependent on grid spacing.

Reply:
I think what you have now works. You may be right that the addition of this absolute term is not necessary, as if `max_change` is very small, then we shouldn't have gotten into this loop in the first place.

Reply:
It definitely is needed though. I tried to get rid of it earlier in the PR, and it was a disaster because it started putting thousands of points in a region that had zero-ish slope in the species equation iirc.